File: dit/inference/counts.c
/**
    An implementation of calculating counts from data.

    This implementation is mostly C90 compatible, but makes use of gcc's
    support for variable-length arrays (VLAs) which was accepted into C99.

    Discussion
    ----------
    windowSize = hLength + fLength
    nWindows = nObservations - windowSize

    In Python, we begin with a list. As we iterate through, we construct tuples
    of the history and future words.  This is O(windowSize). Then, we add
    to the count in the dictionary. This requires a computation of the hashes
    for the history and future words. This is also O(windowSize).  Overall
    then, we are probably at:  O( nWindows * 2 * windowSize ).

    In the C version, there is no need to construct the tuples. Since the data
    is standard, we can construct the hash directly. So this is only
    O(windowSize). Overall then, we are probably at: O(nWindows * windowSize).
    This is a slight improvement.

**/

#include <stdio.h>
#include <math.h>

#include "counts.h"

/**
    Calculates conditional counts of words (single-threaded version)

    Parameters
    ----------
    dataPtr : int *
        A pointer to an array of integer-valued observations.
    nObservations : int
        The number of observations and thus, the length of the data array.
    hLength : int
        The length of the history words.
    fLength : int
        The length of the future words.
    nSymbols : int
        The number of symbols in the alphabet.
    out : int *
        A pointer to an integer array which will contain the counts of future
        words given history words. The array's length should be equal to:
            \frac{nSymbols^{hLength+1} - 1}{nSymbols - 1} * nSymbols^fLength.
    marginalize : int
        If 1, then we marginalize all the counts to obtain conditional counts
        of the future words of length fLength for every history word of length
        0 to length hLength.

    Notes
    -----
    The counts are stored in a flattened row major array.  For each history
    word, the counts of the nSymbols^fLength future words are given by their
    k-ary encoding. For example, if nSymbols=2 and fLength=2, then there are
    four columns representing w=00,01,10,11.  The history words are given by
    their encoding in a k-ary tree.

**/
void counts_st(long* data, int nObservations,
               int hLength, int fLength, int nSymbols,
               long* out, int marginals) {

    int i,j,L,hist;

    /**
        The encoding for histories fixes the right side of the word, as
        that is closest to the cutpoint at t=0. So in breadth-first order,
        the words are:

          \lambda = 0

          0 = 1
          1 = 2

          00 = 3
          10 = 4
          01 = 5
          11 = 6

          000 = 7
          100 = 8
          010 = 9
          110 = 10
          001 = 11
          101 = 12
          011 = 13
          111 = 14

        The encoding for 110, for example, is given by:

            encoding(110) = offset(3) + 1 * 2^0
                                      + 1 * 2^1
                                      + 0 * 2^2
                          = 10
        where

            offset(L) = (nSymbols^L - 1) / (nSymbols - 1)

        When we marginalize the histories, we work with their index only.
        So if we start at w=110, then we want to chop from the left going
        to w=10, w=0, w=\lambda. Of course, we only work with the integers,
        so this corresponds from 10 -> 4, 1, 0.

        The formula describing this progression is:

            w' = offset(L-1) + (w - offset(L)) / nSymbols

        Futures are encoding by appending since the left-side (t=0) of the word
        is fixed. So words of length 2 over a binary alphabet are given the
        following assignments:

            00 -> 0,
            01 -> 1,
            10 -> 2,
            11 -> 3

    **/

    // VLA is allocated on the stack (C99)
    int offset[hLength+2];

    int index, hword, fword;
    int nFutures = pow(nSymbols, fLength);

    if (nSymbols == 1) {
        for (L=0; L < hLength+2; ++L) {
            offset[L] = L;
        }
    }
    else {
        for (L=0; L < hLength+2; ++L) {
            offset[L] = (pow(nSymbols,L) - 1) / (nSymbols - 1);
        }
    }

    /**
        When we parse the data, we only fill in the tail end of `out`, since
        this corresponds to filling out the leaves of the tree.
    **/
    for (i = hLength; i <= nObservations - fLength; ++i) {
        // Construct the index for the history word (from the left)
        hword = offset[hLength];
        L = 0;
        for (j = i - hLength; j < i; ++j) {
            hword += data[j] * pow(nSymbols, L);
            L += 1;
        }
        index = hword * nFutures;

        // Construct the index for the future word (from the right)
        fword = 0;
        L = 0;
        for (j = i + fLength - 1; j >= i; --j) {
            fword += data[j] * pow(nSymbols, L);
            L += 1;
        }
        out[index+fword] += 1;
    }

    if (marginals) {
        // Time-complexity for outer two loops: O( 2^(hLength+1) - 1 )
        int mhist, index_mhist, index_hist;
        for (L = hLength; L > 0; --L) {
            for (hist = offset[L]; hist < offset[L+1]; ++hist) {
                // determine the predecessor history
                mhist = offset[L-1] + (hist - offset[L]) / nSymbols;
                // add the counts for all future words
                index_mhist = mhist * nFutures;
                index_hist = hist * nFutures;
                for (j = 0; j < nFutures; ++j) {
                    out[index_mhist+j] += out[index_hist+j];
                }
            }
        }
    } // end marginalize
}

File: dit/inference/counts.py
"""
Non-cython methods for getting counts and distributions from data.
"""

import contextlib

import numpy as np

__all__ = (
    "counts_from_data",
    "distribution_from_data",
    "get_counts",
)


try:  # cython
    from .pycounts import counts_from_data, distribution_from_data

except ImportError:  # no cython
    from collections import Counter, defaultdict
    from itertools import product

    from boltons.iterutils import windowed_iter

    from .. import modify_outcomes
    from ..exceptions import ditException

    def counts_from_data(data, hLength, fLength, marginals=True, alphabet=None, standardize=True):
        """
        Returns conditional counts from `data`.

        To obtain counts for joint distribution only, use fLength=0.

        Parameters
        ----------
        data : NumPy array
            The data used to calculate morphs. Note: `data` cannot be a generator.
            Also, if standardize is True, then data can be any indexable iterable,
            such as a list or tuple.
        hLength : int
            The maxmimum history word length used to calculate morphs.
        fLength : int
            The length of future words that defines the morph.
        marginals : bool
            If True, then the morphs for all histories words from L=0 to L=hLength
            are calculated.  If False, only histories of length L=hLength are
            calculated.
        alphabet : list
            The alphabet to use when creating the morphs. If `None`, then one is
            obtained from `data`. If not `None`, then the provided alphabet
            supplements what appears in the data.  So the data is always scanned
            through in order to get the proper alphabet.
        standardize : bool
            The algorithm requires that the symbols in data be standardized to
            a canonical alphabet consisting of integers from 0 to k-1, where k
            is the alphabet size.  If `data` is already standard, then an extra
            pass through the data can be avoided by setting `standardize` to
            `False`, but note: if `standardize` is False, then data MUST be a
            NumPy array.


        Returns
        -------
        histories : list
            A list of observed histories, corresponding to the rows in `cCounts`.
        cCounts : NumPy array
            A NumPy array representing conditional counts. The rows correspond to
            the observed histories, so this is sparse. The number of rows in this
            array cannot be known in advance, but the number of columns will be
            equal to the alphabet size raised to the `fLength` power.
        hCounts : NumPy array
            A 1D array representing the count of each history word.
        alphabet : tuple
            The ordered tuple representing the alphabet of the data. If `None`,
            the one is created from the data.


        Notes
        -----
        This requires three complete passes through the data. One to obtain
        the full alphabet. Another to standardize the data.  A final pass to
        obtain the counts.

        This is implemented densely.  So during the course of the algorithm,
        we work with a large array containing a row for each possible history.
        Only the rows corresponding to observed histories are returned.

        """
        with contextlib.suppress(TypeError):
            data = list(map(tuple, data))
        counts = Counter(windowed_iter(data, hLength + fLength))
        cond_counts = defaultdict(lambda: defaultdict(int))
        for word, count in counts.items():
            cond_counts[word[:hLength]][word[hLength:]] += count

        histories = sorted(counts.keys())
        alphabet = set(alphabet) if alphabet is not None else set()
        alphabet = tuple(sorted(alphabet.union(*[set(hist) for hist in histories])))

        cCounts = np.empty((len(histories), len(alphabet) ** fLength))
        for i, hist in enumerate(histories):
            for j, future in enumerate(product(alphabet, repeat=fLength)):
                cCounts[i, j] = cond_counts[hist][future]

        hCounts = cCounts.sum(axis=1)

        return histories, cCounts, hCounts, alphabet

    def distribution_from_data(d, L, trim=True, base=None):
        """
        Returns a distribution over words of length `L` from `d`.

        The returned distribution is the naive estimate of the distribution,
        which assigns probabilities equal to the number of times a particular
        word appeared in the data divided by the total number of times a word
        could have appeared in the data.

        Roughly, it corresponds to the stationary distribution of a maximum
        likelihood estimate of the transition matrix of an (L-1)th order Markov
        chain.

        Parameters
        ----------
        d : list
            A list of symbols to be converted into a distribution.
        L : integer
            The length of the words for the distribution.
        trim : bool
            If true, then words with zero probability are trimmed from the
            distribution.
        base : int or string
            The desired base of the returned distribution. If `None`, then the
            value of `dit.ditParams['base']` is used.

        """
        from dit.distribution import Distribution
        from dit.params import ditParams

        # Normalize data: ensure list, convert 2D rows to tuples for hashability
        d = np.asarray(d)
        d = d.tolist() if d.ndim == 1 else [tuple(row) for row in d]

        if base is None:
            base = ditParams["base"]

        # Build alphabet from all symbols in the data (1D: scalars, 2D: rows)
        alphabet = tuple(sorted(set(d)))

        # Count observed words using sliding windows
        word_counts = Counter(windowed_iter(d, L))

        # Build full distribution over all possible words (including zero-count)
        all_words = list(product(alphabet, repeat=L))
        counts_arr = np.array(
            [word_counts.get(w, 0) for w in all_words],
            dtype=float,
        )
        total = counts_arr.sum()
        pmf = counts_arr / total if total > 0 else np.zeros_like(counts_arr)

        # Flatten nested-tuple words: ((0,), (1,)) -> (0, 1)
        def _flatten_word(w):
            flat = []
            for elem in w:
                if isinstance(elem, tuple) and len(elem) != 1:
                    flat.extend(elem)
                else:
                    flat.append(elem[0] if isinstance(elem, tuple) else elem)
            return tuple(flat)

        words = [_flatten_word(w) for w in all_words]

        # Always build in linear space so zero probabilities are stored as 0.
        # set_base will then correctly convert 0 -> -inf for log bases.
        dist = Distribution(words, pmf, trim=trim, base="linear")

        if L == 1:
            with contextlib.suppress(ditException):
                # Only unwrap 1-tuples (e.g. (0,) -> 0), not multi-variable outcomes (e.g. (0,1,0))
                dist = modify_outcomes(dist, lambda o: o[0] if isinstance(o, tuple) and len(o) == 1 else o)

        # Call set_base after modify_outcomes: modify_outcomes creates a new Distribution
        # with base from the original; if we converted to log first, it would use np.zeros
        # for unfilled positions, giving 0 instead of -inf in log space (which becomes p=1).
        dist.set_base(base)

        return dist


def get_counts(data, length):
    """
    Count the occurrences of all words of `length` in `data`.

    Parameters
    ----------
    data : iterable
        The sequence of samples
    length : int
        The length to group samples into.

    Returns
    -------
    counts : np.array
        Array with the count values.
    """
    hists, _, counts, _ = counts_from_data(data, length, 0)
    mask = np.array([len(h) == length for h in hists])
    counts = counts[mask]
    return counts
